src/Featurizers/Structs.h

```diff
@@ -1,245 +1,329 @@
+            [&numElements, &values](void) -> std::uint64_t & {
+                if(numElements == 0)
+                    throw std::invalid_argument("'numElements' is 0");
+
+                if(numElements < values.size())
+                    throw std::invalid_argument("'numElements' is invalid");
+
+                return numElements;
+            }()
+        )
+    ),
+    Values(
+        std::move(
+            [&values](void) -> std::vector<ValueEncoding> & {
+                if(values.empty())
+                    throw std::invalid_argument("'values' is empty");
+
+                // Ensure that the values are ordered by index
+                std::uint64_t               prevIndex(0);
+
+                for(auto iter = values.cbegin(); iter != values.cend(); ++iter) {
+                    if(iter->Index <= prevIndex && iter != values.cbegin())
+                        throw std::invalid_argument("'values' is not ordered");
+
+                    prevIndex = iter->Index;
+                }
+
+                return values;
+            }()
+        )
+    ) {
+}
+
+template <typename T>
+bool SparseVectorEncoding<T>::operator==(SparseVectorEncoding const &other) const {
+    return NumElements == other.NumElements
+        && Values == other.Values;
+}
+
+template <typename T>
+bool SparseVectorEncoding<T>::operator!=(SparseVectorEncoding const &other) const {
+    return (*this == other) == false;
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  SingleValueSparseVectorEncoding
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+SingleValueSparseVectorEncoding<T>::SingleValueSparseVectorEncoding(std::uint64_t numElements, value_type value, std::uint64_t index) :
+    NumElements(
+        std::move(
+            [&numElements](void) -> std::uint64_t & {
+                if(numElements == 0)
+                    throw std::invalid_argument("'numElements' is 0");
+
+                return numElements;
+            }()
+        )
+    ),
+    Value(std::move(value)),
+    Index(
+        std::move(
+            [this, &index](void) -> std::uint64_t & {
+                if(index >= this->NumElements)
+                    throw std::invalid_argument("'index' is invalid");
+
+                return index;
+            }()
+        )
+    ) {
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#endif
+
+template <typename T>
+bool SingleValueSparseVectorEncoding<T>::operator==(SingleValueSparseVectorEncoding const &other) const {
+    return NumElements == other.NumElements
+        && Value == other.Value
+        && Index == other.Index;
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic pop
+#endif
+
+template <typename T>
+bool SingleValueSparseVectorEncoding<T>::operator!=(SingleValueSparseVectorEncoding const &other) const {
+    return (*this == other) == false;
```

Comment:
> == false; [](start = 27, length = 10)

You compare bool to another bool to get a bool. You have so many bools!
return !(*this == other);